lib/c-api/src/wasm_c_api/engine/mod.rs

```diff
@@ -0,0 +1,298 @@
+            #[cfg(feature = "jsc")]
+            wasmer_api::BackendKind::Jsc => wasmer_engine_t::JSC,
+            // Should be unreachable, however.
+            b => {
+                #[cfg(feature = "sys")]
+                {
+                    #[cfg(feature = "cranelift")]
+                    {
+                        if matches!(b, wasmer_api::BackendKind::Cranelift) {
+                            return wasmer_engine_t::UNIVERSAL;
+                        }
+                    }
+                    #[cfg(feature = "singlepass")]
+                    {
+                        if matches!(b, wasmer_api::BackendKind::Singlepass) {
+                            return wasmer_engine_t::UNIVERSAL;
+                        }
+                    }
+                    #[cfg(feature = "llvm")]
+                    {
+                        if matches!(b, wasmer_api::BackendKind::LLVM) {
+                            return wasmer_engine_t::UNIVERSAL;
+                        }
+                    }
+
+                    if matches!(b, wasmer_api::BackendKind::Headless) {
+                        return wasmer_engine_t::UNIVERSAL;
+                    }
+                }
+
+                panic!("Unsupported backend: {b:?}")
+            }
+        }
+    }
+}
+
+/// An engine is used by the store to drive the compilation and the
+/// execution of a WebAssembly module.
+///
+/// cbindgen:ignore
+#[repr(C)]
+// We can let the API decide which engine is default with the given set of
+// features.
+//
+// See the impl of `Default` for `BackendEngine` in the `wasmer` (API) crate.
+#[derive(Default)]
+pub struct wasm_engine_t {
+    pub(crate) inner: Engine,
+}
+
+#[no_mangle]
+#[allow(unreachable_code)]
+pub extern "C" fn wasm_engine_new() -> Box<wasm_engine_t> {
+    Box::new(wasm_engine_t::default())
+}
+
+/// Deletes an engine.
+///
+/// # Example
+///
+/// ```rust
+/// # use wasmer_inline_c::assert_c;
+/// # fn main() {
+/// #    (assert_c! {
+/// # #include "tests/wasmer.h"
+/// #
+/// int main() {
+///     // Create a default engine.
+///     wasm_engine_t* engine = wasm_engine_new();
+///
+///     // Check we have an engine!
+///     assert(engine);
+///
+///     // Free everything.
+///     wasm_engine_delete(engine);
+///
+///     return 0;
+/// }
+/// #    })
+/// #    .success();
+/// # }
+/// ```
+///
+/// cbindgen:ignore
+#[no_mangle]
+pub unsafe extern "C" fn wasm_engine_delete(_engine: Option<Box<wasm_engine_t>>) {}
+
+/// Creates an engine with a particular configuration.
+///
+/// # Example
+///
+/// See [`wasm_config_new`].
+///
+/// cbindgen:ignore
+#[no_mangle]
+pub extern "C" fn wasm_engine_new_with_config(
+    config: Option<Box<wasm_config_t>>,
+) -> Option<Box<wasm_engine_t>> {
+    #[allow(unused)]
+    let config = *(config?);
+
+    match config.engine {
+        #[cfg(feature = "sys")]
+        wasmer_engine_t::UNIVERSAL => config::sys::wasm_sys_engine_new_with_config(config),
```

Comment:
Should we have the Cranelift / LLVM / Singlepass as well here?

Reply:
Perhaps, to keep symmetry with the (non-c-) api, we should also rename `wasmer_engine_t` to `wasmer_backend_t`.